pkg/apis/build/v1alpha1/buildrun_types.go

```diff
@@ -0,0 +1,84 @@
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	LabelBuildRun              = "buildrun.build.dev/name"
+	LabelBuildRunGeneration    = "buildrun.build.dev/generation"
+)
+
+// BuildRunSpec defines the desired state of BuildRun
+type BuildRunSpec struct {
+
+	// BuildRef refers to the Build
+	BuildRef *BuildRef `json:"buildRef"`
+
+	// ServiceAccount refers to the kubernetes serviceaccount
+	// which is used for resource control.
+	// Default serviceaccount will be set if it is empty
+	// +optional
+	ServiceAccount *string `json:"serviceAccount,omitempty"`
+}
+
+// BuildRunStatus defines the observed state of BuildRun
+type BuildRunStatus struct {
+
+	// The Succeeded status of the TaskRun
+	Succeeded corev1.ConditionStatus `json:"succeeded,omitempty"`
+
+	// The Succeeded reason of the TaskRun
+	Reason string `json:"reason,omitempty"`
+
+	// PodName is the name of the pod responsible for executing this task's steps.
+	LatestTaskRunRef *string `json:"latestTaskRunRef,omitempty"`
```

Comment:
Hi,

Below is I thought,

- First, I think BuildRun can have multiple TaskRun (For example, in future, we can re-run the BuildRun from cli or console, then we can just trigger a new TaskRun, then the BuildRun status will be changed to running and the 
LatestTaskRunRef will be changed too.

- Second, yes, we need ownerRef, and I plan to implement it in another PR by using webhook, when you create the new BuildRun yaml, the controller can help you set the onwerRef automatically.

Of course, such as validate and default, we will continue improving later soon.

Reply:
Thanks for explaining!

I would avoid having the same BuildRun re-used again for multiple runs. 

The way I see BuildRun is that it is a unit of execution. One would need a new BuildRun to have a new build execution.

So if someone wanted to find out "how many builds have run?", the answer would be in "How many BuildRuns are present".

Reply:
OK, it also makes sense.

So I just need to remove this `LatestTaskRunRef` from the status and api. That is enough.

Then it will same as before. If no other suggestion, I will do it today.

Reply:
In future, we need to see how to make the BuildRun CR immutable.

Ideally, I wouldn't want the user to modify the BuildRun in the middle of an execution and cause the generation of a new TaskRun. That would cause us to lose the initial BuildRun's  information ( unless we append conditions?)

I gave this a rethink, let's keep the LatestTaskRunRef as is. We can visit it later :)

Reply:
Read the code again, it seem if that if there's a change in BuildRun.spec, control enters the reconcile loop, finds an existing TaskRun, updates status if needed, and then concludes the reconcile - thereby leaving the buildrun spec in a modified state which didn't really reflect on the TaskRun.


This behaviour is OK for now, too.

Reply:
Yes, the current logic only generate a new taskrun when there is no taskrun in the BuildRun. It won’t regenerate a new taskrun for the BuildRun from the *buildrun controller reconcile logic*